pkg/cache/multi_namespace_cache.go

```diff
@@ -37,39 +39,88 @@ type NewCacheFunc func(config *rest.Config, opts Options) (Cache, error)
 // a new global namespaced cache to handle cluster scoped resources.
 const globalCache = "_cluster-scope"
 
-// MultiNamespacedCacheBuilder - Builder function to create a new multi-namespaced cache.
-// This will scope the cache to a list of namespaces. Listing for all namespaces
-// will list for all the namespaces that this knows about. By default this will create
-// a global cache for cluster scoped resource. Note that this is not intended
-// to be used for excluding namespaces, this is better done via a Predicate. Also note that
-// you may face performance issues when using this with a high number of namespaces.
-func MultiNamespacedCacheBuilder(namespaces []string) NewCacheFunc {
+// ByNamespaceOptions is used to configure the functions used to create caches
+// on a per-namespace basis.
+type ByNamespaceOptions struct {
+	NewNamespaceCaches       map[string]NewCacheFunc
+	NewClusterCache          NewCacheFunc
+	NewDefaultNamespaceCache NewCacheFunc
+}
+
+// BuilderByNamespace builds a composite cache that delegates to per-namespace
+// caches built according to the passed ByNamespaceOptions. If NewDefaultNamespaceCache
+// is defined, it will be used as a catch-all for objects not in the namespaces defined in
+// NewNamespaceCaches. The default namespace cache is automatically configured with extra
+// field selectors to avoid duplicate caching of objects between namespace-specific caches
+// and this catch-all cache. NewClusterCache is used to create a cache for cluster-scoped
+// objects. If it is undefined, a default cache will be created using the New function.
+func BuilderByNamespace(byNamespaceOpts ByNamespaceOptions) NewCacheFunc {
 	return func(config *rest.Config, opts Options) (Cache, error) {
 		opts, err := defaultOpts(config, opts)
 		if err != nil {
 			return nil, err
 		}
 
-		caches := map[string]Cache{}
-
-		// create a cache for cluster scoped resources
-		gCache, err := New(config, opts)
+		if byNamespaceOpts.NewClusterCache == nil {
+			byNamespaceOpts.NewClusterCache = New
+		}
+		clusterCache, err := byNamespaceOpts.NewClusterCache(config, opts)
 		if err != nil {
-			return nil, fmt.Errorf("error creating global cache: %w", err)
+			return nil, err
 		}
 
-		for _, ns := range namespaces {
+		nsToCache := map[string]Cache{}
+		if byNamespaceOpts.NewDefaultNamespaceCache != nil {
+			defaultNamespaceCache, err := byNamespaceOpts.NewDefaultNamespaceCache(config, ignoreNamespaces(opts, byNamespaceOpts.NewNamespaceCaches))
+			if err != nil {
+				return nil, err
+			}
+			nsToCache[corev1.NamespaceAll] = defaultNamespaceCache
```

Comment:
> I thought NewDefaultNamespaceCache was intended to be the default used if a namespace has no explicit NewCacheFunc?

That's correct.

> It looks like we fall through to this cache if no other cache matches, but that is new behavior or not?

That is new behavior, but only if you're using the `ByNamespaceBuilder` variation of this function. The existing `MultiNamespacedCacheBuilder` leaves `NewDefaultNamespaceCache` set to `nil`, meaning there is no catch-all global cache, so the behavior there remains the same.

Reply:
> That is new behavior, but only if you're using the ByNamespaceBuilder variation of the builder. The existing MultiNamespacedCacheBuilder leaves NewDefaultNamespaceCache set to nil, meaning there is no catch-all global cache, so the behavior there remains the same.

IMHO it is extremely unintuitive that this happens in one builder and not the other and IMHO it is wrong to do this by default. If anything, we should have an explicit knob to opt it in to this in both the builders.

Also IMHO rather than forcing a selector for all namespaces by using the map, it would be nicer to have a struct as value that optionally takes a selector and an option for a `DefaultSelector` that if set will be used if a namespace does not have a selector.

Reply:
> IMHO it is extremely unintuitive that this happens in one builder and not the other and IMHO it is wrong to do this by default. If anything, we should have an explicit knob to opt it in to this in both the builders.

- In the `BuilderByNamespace`, it is opt-in. You have to specifically set `NewDefaultNamespaceCache`.
- In `MultiNamespacedCacheBuilder` there has never been a fall-back option, and my thought was to not make a breaking change to that builder whatsoever, meaning there's no way to opt-in.

> Also IMHO rather than forcing a selector for all namespaces by using the map, it would be nicer to have a struct as value that optionally takes a selector and an option for a DefaultSelector that if set will be used if a namespace does not have a selector.

I'm not sure I'm following this. There's nothing in this layer that deals directly with selectors. This is just "I want different `NewCacheFunc`s per specific namespace/catch-all namespaces/cluster-scope."